File: Tests/Advection/Advection.cpp
#include <iostream>

#include <lsAdvect.hpp>
#include <lsDomain.hpp>
#include <lsExpand.hpp>
#include <lsMakeGeometry.hpp>
#include <lsPrune.hpp>
#include <lsTestAsserts.hpp>
#include <lsToMesh.hpp>
#include <lsToSurfaceMesh.hpp>
#include <lsVTKWriter.hpp>

/**
  This example shows how to use lsAdvect to create an egg
  shape from a spherical level set using directional growth rates.
  \example Advection.cpp
*/

// implement own velocity field
class velocityField : public lsVelocityField<double> {
public:
  double getScalarVelocity(const std::array<double, 3> & /*coordinate*/,
                           int /*material*/,
                           const std::array<double, 3> &normalVector,
                           unsigned long /*pointId*/) final {
    // Some arbitrary velocity function of your liking
    // (try changing it and see what happens :)
    double velocity = 1. + ((normalVector[0] > 0) ? 2.3 : 0.5) *
                               std::abs(normalVector[0] * normalVector[0]);
    return velocity;
  }

  std::array<double, 3>
  getVectorVelocity(const std::array<double, 3> & /*coordinate*/,
                    int /*material*/,
                    const std::array<double, 3> & /*normalVector*/,
                    unsigned long /*pointId*/) final {
    return std::array<double, 3>({});
  }
};

int main() {

  constexpr int D = 3;
  omp_set_num_threads(4);

  double gridDelta = 0.4999999;

  auto sphere1 = lsSmartPointer<lsDomain<double, D>>::New(gridDelta);

  double origin[3] = {5., 0., 0.};
  double radius = 7.3;

  lsMakeGeometry<double, D>(
      sphere1, lsSmartPointer<lsSphere<double, D>>::New(origin, radius))
      .apply();

  // {
  //   std::cout << "Extracting..." << std::endl;
  //   auto mesh = lsSmartPointer<lsMesh<>>::New();
  //   lsToSurfaceMesh<double, D>(sphere1, mesh).apply();
  //   lsVTKWriter<double>(mesh, "before.vtk").apply();
  // }

  // instantiate velocities
  auto velocities = lsSmartPointer<velocityField>::New();

  // std::cout << "Advecting" << std::endl;

  lsAdvect<double, D> advectionKernel;
  advectionKernel.insertNextLevelSet(sphere1);
  advectionKernel.setVelocityField(velocities);
  advectionKernel.setIntegrationScheme(
      lsIntegrationSchemeEnum::ENGQUIST_OSHER_1ST_ORDER);
  advectionKernel.setSaveAdvectionVelocities(true);

  double time = 0.;
  for (unsigned i = 0; time < 2.0 && i < 1e3; ++i) {
    // auto mesh = lsSmartPointer<lsMesh<>>::New();
    // lsToMesh<double, D>(sphere1, mesh).apply();
    // // mesh->print();
    // lsVTKWriter<double>(mesh, "after-" + std::to_string(i) + ".vtk").apply();

    advectionKernel.apply();
    time += advectionKernel.getAdvectedTime();
  }

  LSTEST_ASSERT_VALID_LS(sphere1, double, D)

  // std::cout << sphere1->getNumberOfPoints() << std::endl;

  // lsPrune<double, D>(sphere1).apply();
  // lsExpand<double, D>(sphere1, 2).apply();

  // {
  //   std::cout << "Extracting..." << std::endl;
  //   auto mesh = lsSmartPointer<lsMesh<>>::New();
  //   lsToSurfaceMesh<double, D>(sphere1, mesh).apply();
  //   mesh->print();
  //   lsVTKWriter<double>(mesh, "after.vtk").apply();
  // }

  return 0;
}

File: tests/Advection/Advection.py
import viennals as vls

vls.setDimension(3)

# Implement own velocity field
class VelocityField(vls.VelocityField):
    def __init__(self):
        super().__init__()

    def getScalarVelocity(self, coord, material, normal, pointId):
        # Some arbitrary velocity function
        velocity = 1.0 + (2.3 if normal[0] > 0 else 0.5) * abs(normal[0] * normal[0])
        return velocity

    def getVectorVelocity(self, coord, material, normal, pointId):
        return [0.0, 0.0, 0.0]

def main():
    # Set number of threads
    vls.setNumThreads(8)

    gridDelta = 0.6999999

    discretizationSchemes = [
        vls.DiscretizationSchemeEnum.ENGQUIST_OSHER_1ST_ORDER,
        vls.DiscretizationSchemeEnum.ENGQUIST_OSHER_2ND_ORDER,
        vls.DiscretizationSchemeEnum.LAX_FRIEDRICHS_1ST_ORDER,
        vls.DiscretizationSchemeEnum.LAX_FRIEDRICHS_2ND_ORDER,
        vls.DiscretizationSchemeEnum.LOCAL_LOCAL_LAX_FRIEDRICHS_1ST_ORDER,
        vls.DiscretizationSchemeEnum.LOCAL_LOCAL_LAX_FRIEDRICHS_2ND_ORDER,
        vls.DiscretizationSchemeEnum.LOCAL_LAX_FRIEDRICHS_1ST_ORDER,
        vls.DiscretizationSchemeEnum.LOCAL_LAX_FRIEDRICHS_2ND_ORDER,
        vls.DiscretizationSchemeEnum.WENO_5TH_ORDER
    ]

    for scheme in discretizationSchemes:
        sphere1 = vls.Domain(gridDelta)

        origin = [5.0, 0.0, 0.0]
        radius = 7.3

        vls.MakeGeometry(sphere1, vls.Sphere(origin, radius)).apply()

        # Instantiate velocities
        velocities = VelocityField()

        advectionKernel = vls.AdvectRungeKutta3()
        advectionKernel.insertNextLevelSet(sphere1)
        advectionKernel.setVelocityField(velocities)
        advectionKernel.setDiscretizationScheme(scheme)
        advectionKernel.setSaveAdvectionVelocities(True)

        time = 0.0
        i = 0
        while time < 1.0 and i < 100:
            advectionKernel.apply()
            time += advectionKernel.getAdvectedTime()

            fileName = f"{scheme}_{i}.vtp"
            mesh = vls.Mesh()
            vls.ToMesh(sphere1, mesh).apply()
            vls.VTKWriter(mesh, "points_" + fileName).apply()
            vls.ToSurfaceMesh(sphere1, mesh).apply()
            vls.VTKWriter(mesh, "surface_" + fileName).apply()
            i += 1
        print(f"Done scheme {scheme}")

if __name__ == "__main__":
    main()
